### PR TITLE
Improve the speed of `DOMQuery.remove`

### PR DESCRIPTION
**Pull request by davep**
See https://github.com/Textualize/textual/discussions/1110 for some context. In short, this changes the `Remove` event so that it can take a collection of widgets to remove, rather than a single widget. The way that removal is done is then changed to dismantle the relevant parts of the DOM first, perform the closedown of the widgets involved, and then finally do a single refresh operation.

Note that, along the way, what was `DOMNode.ancestors` has become `DOMNode.ancestors_with_self` and `DOMNode.ancestors` now no longer contains `self`.

### CODE REVIEW

File: src/textual/dom.py [hunk 1]
```diff
@@ -561,6 +564,12 @@ def ancestors(self) -> list[DOMNode]:
             node = node._parent
         return cast("list[DOMNode]", nodes)
 
+    @property
+    def ancestors(self) -> list[DOMNode]:
+        """list[DOMNode]: A list of ancestor nodes Nodes by tracing ancestors all the way back to App."""
+        nodes = self.ancestors_with_self
+        return nodes[1:] if nodes else nodes
```
Comment (willmcgugan) [inline src/textual/dom.py:571]: I think we can guarantee that nodes has at least one node.

File: src/textual/app.py [hunk 2]
```diff
@@ -1822,23 +1824,66 @@ async def _on_resize(self, event: events.Resize) -> None:
...
+        # and reduce it to the minimal set of things that need removing. In
+        # other words, for any given branch that is going to have things
+        # removed, we need to just keep the topmost node in the DOM that is
+        # a candidate for removal. Note that we retain the ordering of the
+        # original walk at the top as it's depth-first and so should be the
+        # order we want to go in from here on.
+        pruned_remove: list[Widget] = [
+            widget
+            for widget in everything_to_remove
+            if not any(
```
Comment (willmcgugan) [inline src/textual/app.py:1866]: I think we can replace the if condition with a set operation:

```python
isdisjoint = everything_to_remove
pruned_remove = [
    widget for widget in everything_to_remove
    if isdisjoint(widget.ancestors)
]
```

File: src/textual/app.py [hunk 3]
```diff
@@ -1822,23 +1824,66 @@ async def _on_resize(self, event: events.Resize) -> None:
...
+        ]
+
+        # Now that we know that minimal set of widgets, we go through them
+        # and get their parents to forget about them. This has the effect of
+        # snipping each affected branch from the DOM.
+        for widget in pruned_remove:
+            if widget.parent is not None:
+                widget.parent.children._remove(widget)
+
+        # Havnig done that, it's now safe for us to start the process of
```
Comment (willmcgugan) [inline src/textual/app.py:1878]:
```suggestion
        # Having done that, it's now safe for us to start the process of
```

**Review by willmcgugan — CHANGES_REQUESTED** [review]
Great. Suggested an optimization.

File: src/textual/dom.py (hunk 1 above)
Comment (davep) [reply src/textual/dom.py:571]: Good point. :-D

File: src/textual/app.py (hunk 2 above)
Comment (davep) [reply src/textual/app.py]: Like this? https://github.com/Textualize/textual/pull/1147/commits/3cb00f93d5e2b000f2fa19951b5338ac0ee59180

File: src/textual/app.py [hunk 4]
```diff
@@ -1822,23 +1824,64 @@ async def _on_resize(self, event: events.Resize) -> None:
...
+        # Next, we want to go through the full set of everything to remove
+        # and reduce it to the minimal set of things that need removing. In
+        # other words, for any given branch that is going to have things
+        # removed, we need to just keep the topmost node in the DOM that is
+        # a candidate for removal. Note that we retain the ordering of the
+        # original walk at the top as it's depth-first and so should be the
+        # order we want to go in from here on.
+        pruned_remove = [
+            widget
+            for widget in everything_to_remove
```
Comment (willmcgugan) [inline src/textual/app.py:1865]: I have a feeling we are doing more work than required here. I think we could get the same result by iterating over `event.widgets`.

Comment (davep) [reply src/textual/app.py]: I think you're right, good catch.

Comment (davep) [reply src/textual/app.py]: Done: a888e52c30191b17e1d885b64f3a0e2b61faacdf

**Review by willmcgugan — APPROVED** [review]

**Pull request merged**